packages/labs/src/inputList.tsx

```diff
@@ -0,0 +1,293 @@
+            }
+        }
+    }
+
+    private getActiveElement() {
+        if (this.itemsParentRef != null) {
+            return this.itemsParentRef.children.item(this.getActiveIndex()) as HTMLElement;
+        }
+        return undefined;
+    }
+
+    private getActiveIndex() {
+        // NOTE: this operation is O(n) so it should be avoided in render(). safe for events though.
+        return this.state.filteredItems.indexOf(this.props.activeItem);
+    }
+
+    private getItemsParentPadding() {
+        const { paddingTop, paddingBottom } = getComputedStyle(this.itemsParentRef);
+        return {
+            paddingBottom: pxToNumber(paddingBottom),
+            paddingTop: pxToNumber(paddingTop),
+        };
+    }
+
+    private handleItemSelect = (item: T, event: React.SyntheticEvent<HTMLElement>) => {
+        Utils.safeInvoke(this.props.onActiveItemChange, item);
+        Utils.safeInvoke(this.props.onItemSelect, item, event);
+    }
+
+    private handleKeyDown = (event: React.KeyboardEvent<HTMLElement>) => {
+        switch (event.keyCode) {
+            case Keys.ARROW_UP:
+                event.preventDefault();
+                this.moveActiveIndex(-1);
+                break;
+            case Keys.ARROW_DOWN:
+                event.preventDefault();
+                this.moveActiveIndex(1);
+                break;
+            default: return;
+        }
+        Utils.safeInvoke(this.props.onKeyDown, event);
+    }
+
+    private handleKeyUp = (event: React.KeyboardEvent<HTMLElement>) => {
+        const { activeItem, onItemSelect, onKeyUp } = this.props;
+        // using keyup for enter to play nice with Button's keyboard clicking.
+        // if we were to process enter on keydown, then Button would click itself on keyup
+        // and the popvoer would re-open out of our control :(.
+        if (event.keyCode === Keys.ENTER) {
+            event.preventDefault();
+            Utils.safeInvoke(onItemSelect, activeItem, event);
+        }
+        Utils.safeInvoke(onKeyUp, event);
+    }
+
+    private moveActiveIndex(direction: number) {
+        // indicate that the active item may need to be scrolled into view after update.
+        // this is not possible with mouse hover cuz you can't hover on something off screen.
+        this.shouldCheckActiveItemInViewport = true;
+        const { filteredItems } = this.state;
+        const maxIndex = Math.max(filteredItems.length - 1, 0);
+        const nextActiveIndex = Utils.clamp(this.getActiveIndex() + direction, 0, maxIndex);
+        Utils.safeInvoke(this.props.onActiveItemChange, filteredItems[nextActiveIndex]);
+    }
+}
+
+function pxToNumber(value: string) {
+    return parseInt(value.slice(0, -2), 10);
+}
+
+function getFilteredItems<T>({ items, itemPredicate, itemListPredicate, query }: IInputListProps<T>) {
+    if (Utils.isFunction(itemListPredicate)) {
+        // TODO: this can reorder the items, which can cause weirdness while filtering. do some design.
```

Comment:
turns out this one is fine--the reordering is useful and the component handles it well.